### ISSUE TITLE
Introduce strict boolean assertions

### ISSUE THREAD

**Issue by ventuno** [qunitjs/qunit#1444]
[`assert.ok`](https://api.qunitjs.com/assert/ok)/[`assert.notOk`](https://api.qunitjs.com/assert/notOk) pass when their argument is truthy/falsy.
They definitely have their usage but are also potentially error prone (see [`no-assert-ok`](https://github.com/platinumazure/eslint-plugin-qunit/blob/master/docs/rules/no-assert-ok.md) in [`eslint-plugin-quinit`](https://github.com/platinumazure/eslint-plugin-qunit) for a full discussion) and using strict equality (e.g.: `assert.strictEqual(... true/false)`) should be preferable as we should write tests that assert against exact values as much as possible.

A lot of developers who understand the potential issues with these assertions still prefer using them because of their ergonomics. It is indeed much nicer to read:
`assert.ok(myFunctionThatShouldReturnTrue())` than `assert.strictEqual(myFunctionThatShouldReturnTrue(), true)` (it's also much shorter and we all know how much developers love saving inputs).

I'm thus proposing the introduction of a new set of assertions: `strictlyTrue`/`strictlyFalse` that similarly to `strictEqual` will check for strict equality (`===`) of their input against `true`/`false`.

### Sample implementation
https://github.com/qunitjs/qunit/pull/1445

**Comment by trentmwillis** [comment]
I agree that your proposed API reads better, but I'm also not entirely sure it's worth adding as a core assertion since it would be relatively to implement as a custom assertion/plugin.

Any specific reason you feel this should be added to QUnit itself instead of as an extension?

**Comment by ventuno** [comment]
Thanks @trentmwillis. I agree that it would be relatively easy to make it a custom assertion and I don't think I have a much stronger argument than the one I tried to build above.
Let me try to add a few points:
1) similarly to the `assert.equal`/`assert.strictEqual` duality, it makes sense to have a stricter version of the `assert.ok` check;
2) in general it's pretty easy to build the argument to prefer `strictEqual` instead of `equal` (just like we know `===` is better than `==`), but much harder to get people to use `strictEqual(..., true)` instead of `ok(...)`, having `strictlyEqual` would make that argument a lot easier;
3) other JavaScript assertion libraries like Chai support [`assert.isTrue`](https://www.chaijs.com/api/assert/#method_istrue) together with [`assert.ok`](https://www.chaijs.com/api/assert/#method_isok).

**Comment by trentmwillis** [comment]
I think I see the value in this. @qunitjs/core anyone else have an opinion on whether this should be a core feature?

**Comment by Krinkle** [comment]
[I agree (2015 blogpost)](https://timotijhof.net/posts/2015/qunit-anti-patterns/) on discouraging use of `assert.ok()`, and I also agree that it's not obvious to users today that this is a bad API, or what they should use instead.

That kind of problem isn't always best solved with additional assertion methods. But, it might be in this case.

> A lot of developers who understand the potential issues with these assertions still prefer using them because of their ergonomics. It is indeed much nicer to read:
`assert.ok(myFunctionThatShouldReturnTrue())` than `assert.strictEqual(myFunctionThatShouldReturnTrue(), true)` (it's also much shorter and we all know how much developers love saving inputs).

I get 100% that the current issue needs to be improved, and finding `ok()` in the docs today looks appealing, and needing to change one's code doens't always feel right. I also get that if `isTrue()` were in the list of assertion methods, one would have likely selected that instead, thus avoiding the issue entirely. But, I'm not sure that's enough of a reason by itself as it doesn't scale in my opinion.

I'd like to push back on the ergonomics point for a bit. If you mean that it's more typing, I'd say I believe engineers spend far less time typing actual source code than one might think and it's a common fallacy to think shortening identifiers or keywords helps anyone in the long term. Having said that, QUnit is built on simplicity and I believe that it does matter a lot to reduce the number of concepts required to write, review or debug code. E.g. If the same amount of intent, logic, and responsibility can be expressed with fewer entities in a statement, that's imho a good thing.

In the case of `assert.isTrue value` vs `asssert.strictEquals value true`, I'm not sure if it adds an abstraction. In my mind the phrase `isTrue` still very much means `equalling the value of true`, it isn't something we can hide.

It does mean fewer parameters specified. As such, if assertions could generally be decoupled from the expected output values, or if there were only a limited set of outputs possible, then such an abstraction could allow users to sometimes never need to worry about that additional aspect.

For example, if an API can only respond with one of three JSON messages, but there are dozens of test cases, one might want abstract those behind a constant (`assert.strictEqual output MSG_FOO`), or an assertion helper (`assert.msgIsFoo output`). However for test suites more generally, I don't think we can consider the `actual` parameter as something that is avoidable or redunant.

```js
assert.strictEquals( …, null, 'the default' );
assert.strictEquals( …, 'f' );
assert.strictEquals( …, 'fooo' );
assert.strictEquals( …, false, 'more than three Os' );
assert.strictEquals( …, 'b' );
```

Compared to

```js
assert.strictEquals( …, null, 'the default' );
assert.strictEquals( …, 'f' );
assert.strictEquals( …, 'fooo' );
assert.isFalse( …, 'more than three Os' );
assert.strictEquals( …, 'b' );
```

Note that this would be the first assertion shortcut we'd ship with QUnit. That's not a blocker by any means, though!

I think if isTrue/isFalse methods were to exist, they should be simple and always strict, e.g. no strict/loose options or variants. 

@ventuno What do you think about the above? Would a recommendation in the `assert.ok()` documentation have helped for users that don't use the eslint plugin? Perhaps you could share a few words about what aspect of ergonomics stands out for you, e.g. what decision or concept it would help skip. I think there's potential here, but I'd like to end it with rationale we can build on more :)

**Comment by ventuno** [comment]
Thanks a lot @Krinkle for the thoughtful response.

> Would a recommendation in the `assert.ok()` documentation have helped for users that don't use the eslint plugin?

In general, I think that adding something to the official documentation that discourages the use of `assert.*ok` would help greatly to point people towards the right direction. If that's what you're suggesting, this should probably done regardless.

> Perhaps you could share a few words about what aspect of ergonomics stands out for you [...]

When it comes to ergonomics, my reasoning stems from three ideas:
1) it's easier to read. Right now whenever I see `assert.ok(...)` it's automatic for me to assume that I'm expecting _something_ to be `true`. When I see `assert.strictEqual` I always need to read the entire expression to know what the expected return value is (which I agree is not necessarily a bad thing);

2) people "expect" (prefer?) not to specify a second argument in boolean expressions if the type is boolean. Let's consider the example of typed languages (e.g.: Java), in which you are expected to write:
```Java
final String someString = "aString";
if (someString === "aString") {
  // ...
}
```
but you'll never see (although it's valid):
```Java
final boolean someBoolean = true;
if (someBoolean === true) {
  // ...
}
```
and instead write:
```Java
final boolean someBoolean = true;
if (someBoolean) {
  // ...
}
```
so for me it's totally legit to expect a "special" assertion for the boolean case.

3) similarly to `assert.equal`/`assert.strictEqual` I would consider `assert.isTrue` the "strict" version of `assert.ok`.

Does that help clarify my point of you?

**Comment by Krinkle** [comment]
@ventuno Yep, I think that makes sense. Another reason is that I think it helps to take away to need to decide between equal and strict equals, and focus purely on boolean true/false. That seems to take away a small decision/burden.

As for naming, I'd prefer something different then `strictlyTrue`. I have two alternate ideas to put forward.

### assert.isTrue / assert.isFalse

This is what some other assertion libaries use, including [Chai](https://www.chaijs.com/api/assert/) (JS), [Spring Framwork](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/util/Assert.html) (Java), and [NUnit](https://docs.microsoft.com/en-us/dotnet/api/nunit.framework.assert.istrue?view=xamarin-ios-sdk-12) (dotnet).

```js
QUnit.test( 'example', assert => {
  assert.strictEquals( obj.x, 'foo' );
  assert.propEquals( obj.y, { foo: true } );
  assert.isTrue( obj.y.foo );
  assert.isFalse( obj.y.bar );
} );
```

Note that the aforementioned libraries have a tradition of using verbs in front methods, whereas QUnit currently does not.

### assert.true / assert.false

This might stylistically feel more at home in the JavaScript language and in QUnit.

```js
QUnit.test( 'example', assert => {
  assert.strictEquals( obj.x, 'foo' );
  assert.propEquals( obj.y, { foo: true } );
  assert.true( obj.y.foo );
  assert.false( obj.y.bar );
} );
```

Thoughts?

**Comment by ventuno** [comment]
Thanks @Krinkle. I also like `assert.true`/`assert.false` better. I updated https://github.com/qunitjs/qunit/issues/1444 to match the new names.

I'm not sure what I was thinking when I picked `strictlyTrue`/`strictlyFalse`, but I wasn't too happy with that either. I probably wanted to convey some similarity with `strictEqual`, hinting at the fact that this would be a _strict_ boolean check. In this case then name should have been `strictTrue`/`strictFalse` anyways.

Let me know how you'd like to proceed.
